diff -Nrup linux-2.6.20.1-orig/drivers/i2c/busses/i2c-pxa.c linux-2.6.20.1/drivers/i2c/busses/i2c-pxa.c
--- linux-2.6.20.1-orig/drivers/i2c/busses/i2c-pxa.c	2007-02-20 07:34:32.000000000 +0100
+++ linux-2.6.20.1/drivers/i2c/busses/i2c-pxa.c	2007-02-27 16:42:52.000000000 +0100
@@ -822,9 +822,7 @@ static struct pxa_i2c i2c_pxa = {
 static int i2c_pxa_probe(struct platform_device *dev)
 {
 	struct pxa_i2c *i2c = &i2c_pxa;
-#ifdef CONFIG_I2C_PXA_SLAVE
 	struct i2c_pxa_platform_data *plat = dev->dev.platform_data;
-#endif
 	int ret;
 
 #ifdef CONFIG_PXA27x
@@ -853,6 +851,10 @@ static int i2c_pxa_probe(struct platform
 	i2c->adap.algo_data = i2c;
 	i2c->adap.dev.parent = &dev->dev;
 
+	if (plat) {
+		i2c->adap.class = plat->class;
+	}
+	
 	ret = i2c_add_adapter(&i2c->adap);
 	if (ret < 0) {
 		printk(KERN_INFO "I2C: Failed to add bus\n");
diff -Nrup linux-2.6.20.1-orig/include/asm-arm/arch-pxa/i2c.h linux-2.6.20.1/include/asm-arm/arch-pxa/i2c.h
--- linux-2.6.20.1-orig/include/asm-arm/arch-pxa/i2c.h	2007-02-20 07:34:32.000000000 +0100
+++ linux-2.6.20.1/include/asm-arm/arch-pxa/i2c.h	2007-02-27 16:41:47.000000000 +0100
@@ -64,6 +64,7 @@ struct i2c_slave_client;
 struct i2c_pxa_platform_data {
 	unsigned int		slave_addr;
 	struct i2c_slave_client	*slave;
+	unsigned int		class;
 };
 
 extern void pxa_set_i2c_info(struct i2c_pxa_platform_data *info);
