diff -Nru linux-2.6.20.1-orig/drivers/hwmon/adm1021.c linux-2.6.20.1/drivers/hwmon/adm1021.c
--- linux-2.6.20.1-orig/drivers/hwmon/adm1021.c	2007-06-06 11:00:55.975826000 +0200
+++ linux-2.6.20.1/drivers/hwmon/adm1021.c	2007-06-06 12:26:43.712871164 +0200
@@ -124,6 +124,8 @@
 /* (amalysh) read only mode, otherwise any limit's writing confuse BIOS */
 static int read_only;
 
+/** RF-PCB max1617a chip disabled by default */
+static int rfpcb = 0;
 
 /* This is the driver that will be inserted */
 static struct i2c_driver adm1021_driver = {
@@ -270,7 +272,15 @@
 	}
 
 	if (kind == max1617) {
-		type_name = "max1617";
+                /*
+                 * Ales Bardorfer <ales@i-tech.si>:
+                 *
+                 * Functionally, the max1617 and max1617a chips are the same,
+                 * thus we will make them announce the same to
+                 * the outside world (/proc & /sys).
+                 */
+                //type_name = "max1617";
+                type_name = "max1617a";
 	} else if (kind == max1617a) {
 		type_name = "max1617a";
 	} else if (kind == adm1021) {
@@ -287,6 +297,24 @@
 		type_name = "mc1066";
 	}
 
+        /*
+         * Ales Bardorfer <ales@i-tech.si>
+         *
+         * Switch the RF PCB temp sersor off (software standby).
+         */
+#define LIBERA_RF_PCB_ADDR   24
+#define LIBERA_RF_PCB_ID      0
+        if ( (rfpcb == 0)  && ((kind == max1617) || (kind == max1617a)) ) {
+            if ( ( LIBERA_RF_PCB_ID == i2c_adapter_id(adapter) ) &&
+                 ( LIBERA_RF_PCB_ADDR == address ) ) {
+                /* Disable ADC and enable suspend mode */
+                adm1021_write_value(new_client, ADM1021_REG_CONFIG_W, 0x40);
+                kfree(data);
+                printk("adm1021: RF-PCB max1617a chip disabled.\n");
+                return 0;
+            }
+        }
+
 	/* Fill in the remaining client fields and put it into the global list */
 	strlcpy(new_client->name, type_name, I2C_NAME_SIZE);
 	data->type = kind;
@@ -405,12 +433,17 @@
 }
 
 MODULE_AUTHOR ("Frodo Looijaard <frodol@dds.nl> and "
-		"Philip Edelbrock <phil@netroedge.com>");
+		"Philip Edelbrock <phil@netroedge.com>"
+               "Libera RF-PCB max1617a chip disabled by "
+               "Ales Bardorfer <ales@i-tech.si>.");
 MODULE_DESCRIPTION("adm1021 driver");
 MODULE_LICENSE("GPL");
 
 module_param(read_only, bool, 0);
 MODULE_PARM_DESC(read_only, "Don't set any values, read only mode");
 
+module_param(rfpcb, bool, 0);
+MODULE_PARM_DESC (rfpcb, "Enable RF-PCB max1617a chip (default: disabled)");
+
 module_init(sensors_adm1021_init)
 module_exit(sensors_adm1021_exit)
